examples/ui/toolbar.rs

```diff
@@ -0,0 +1,257 @@
+//! This example shows a more advanced ui use case of managing a shared state via a set of button
+//! components.
+
+use bevy::ui::Display;
+use bevy::{prelude::*, winit::WinitSettings};
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+struct ToolbarState {
+    selected: ButtonTag,
+}
+
+impl Default for ToolbarState {
+    fn default() -> Self {
+        Self {
+            selected: ButtonTag::Stop,
+        }
+    }
+}
+
+struct ToolbarEvent {
+    tag: ButtonTag,
+    message: Option<String>,
+    interaction: Interaction,
+}
+
+#[derive(Component, Clone, Hash, Debug, Eq, PartialEq)]
+enum ButtonTag {
+    Play,
+    Pause,
+    Stop,
+}
+
+#[derive(Component)]
+struct ToolbarMessage;
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins)
+        // Only run the app when there is user input. This will significantly reduce CPU/GPU use.
+        .insert_resource(WinitSettings::desktop_app())
+        .add_startup_system(setup)
+        .add_state(ToolbarState::default())
+        .add_event::<ToolbarEvent>()
+        // EventReader systems added before event writer
+        .add_system(toolbar_state_system)
+        .add_system(toolbar_message_system)
+        .add_system(toolbar_style_system)
+        // EventWriter system after EventReaders to prevent ordering issues.
+        .add_system(button_interaction_system)
```

Comment:
> I'm on Mac as well (M1 Max) and it runs fine for me. Although, I did notice that click events happen on mouse down? I'm pretty confident that's a Bevy issue, though.

I changed the ordering of the events and added some comments describing my reasoning and everything feels a lot faster.  It could have just been an order of execution / next frame issue (although I am still getting very slow responses in the `game_menu` example).